File: k-mean/k-mean.py
import numpy as np
from matplotlib import pyplot as plt
from matplotlib import animation

def readInput(inputFilename):
	data = np.loadtxt(inputFilename, delimiter=',')
	return data

def euclideDistance(x1, x2, n):
	d = (x1 - x2).reshape([1, n])
	A = d.dot(d.T)
	return np.sqrt(A[0,0])

class KMean:
	def __init__(self, inputFilename, n, k):
		self.n = n
		self.k = k
		self.X = readInput(inputFilename)
		self.m = self.X.shape[0]
		self.Y = np.zeros(self.m, dtype = np.int)
		self.centroids = np.zeros([self.k, self.n])

	def isConverge(self, old_Y):
		return np.array_equal(self.Y, old_Y)

	def visualize(self, step=None):
		if step is None:
			fig = plt.figure('Clustering result by K-Mean')
		else:
			fig = plt.figure('Step {0}'.format(step))
		ax = plt.subplot(111)
		color = ['r', 'g', 'b']
		for i in range(self.m):
			ax.scatter(self.X[i,0], self.X[i,1], c=color[int(self.Y[i])])

		for i in range(self.k):
			ax.scatter(self.centroids[i,0], self.centroids[i,1], marker='x', c=color[i], s=150)
		plt.show()

	def train(self):
		distance = np.zeros(self.k)
		for i in range(self.k):
			#self.centroids[i] = self.X[np.random.randint(0, self.m)]
			self.centroids[i] = self.X[i]

		iStep = 0
		while True:
			print('Step %d' %(iStep))
			print(self.centroids)
			#self.visualize(iStep)
			
			old_Y = self.Y.copy()
			cnt = np.zeros(self.k)
			for i in range(self.m):
				for j in range(self.k):
					distance[j] = euclideDistance(self.X[i], self.centroids[j], self.n)
				self.Y[i] = np.argmin(distance)

			self.printResult()
			print()

			self.centroids = np.zeros([self.k, self.n])
			for i in range(self.m):
				c = self.Y[i]
				self.centroids[c] += self.X[i]
				cnt[c] += 1
			for i in range(self.k):
				for j in range(self.n):
					self.centroids[i,j] /= cnt[i]

			if self.isConverge(old_Y):
				break
			iStep += 1

	def printResult(self):
		for i in range(self.m):
			print('(', end='')
			for j in range(self.n):
				print('%d' % (self.X[i,j]), end=' ')
			print('): %d', self.Y[i])

if __name__ == '__main__':
	model = KMean('input.txt', 2, 3)
	model.train()
	#model.printResult()
	model.visualize()

File: k-mean/k-mean.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <cstring>
#include <string>
#include <cstdlib>
#include <ctime>
#include <cstdio>
#include <cassert>
using namespace std;

#define MAX 1000
const char DELIMITER[2] = ",";

// Declaration
class KMean
{
private:
	int m;
	int n;
	int k;

	vector<vector<double> > data;
	vector<vector<double> > centroids;
	vector<int> label;
	vector<int> cnt;
	int nSteps;

    void getData(char* datafile);
    void vectorize(vector<vector<string> > v);

    bool isConverge(vector<int> oldLabel);

public:
	KMean();
	KMean(char* datafile, int m, int n, int k);
	void train();
	void printResult();
	void printResult1();
	int assign(vector<double> p);
	//vector<int> getResult();
	~KMean();
};

double euclidean(vector<double> p1, vector<double> p2);
vector<string> tokenize(char* s);
int argmin(vector<double> v);
//string toString(char* s);
// End of declaration

int main()
{
	KMean model("input.txt", 30, 2, 3);
	model.train();
	freopen("out.csv", "w", stdout);
	model.printResult1();
	fclose(stdout);
}

// Implementation
KMean::KMean(char* datafile, int m, int n, int k) : m(m), n(n), k(k)
{
	data.resize(m);
	label.resize(m);
	for(int i = 0; i < m; ++i)
	{
		data[i].resize(n);
		label[i] = 0;
	}

	centroids.resize(k);
	cnt.resize(k);
	for(int i = 0; i < k; ++i)
	{
		centroids[i].resize(n);
		cnt[i] = 0;
	}

	getData(datafile);
	nSteps = 0;
}

void KMean::getData(char* datafile)
{
	char s[100];
	vector<string> tmp;
	FILE* f =  fopen(datafile, "r");

	for(int i = 0; i < m; ++i)
	{
		fgets(s, MAX, f);
		tmp = tokenize(s);
		for(int j = 0; j < tmp.size(); ++j)
			data[i][j] = atof(tmp[j].c_str());
	}
    fclose(f);
}

bool KMean::isConverge(vector<int> oldLabel)
{
	bool IsConverge = true;
	for(int i = 0; i < m; ++i)
		if (label[i] != oldLabel[i])
		{
			IsConverge = false;
			break;
		}
	return IsConverge;
}

int KMean::assign(vector<double> p)
{
	int Label;
	vector<double> distance(k, 0.0);
	for(int j = 0; j < k; ++j)
	{
		distance[j] = euclidean(p, centroids[j]);
	}
	Label = argmin(distance);
	return Label;
}

void KMean::train()
{
	srand(time(NULL));
	for(int i = 0; i < k; ++i)
	{
		int idx = ((rand() % m) + (rand() % m)) % m;
		centroids[i] = data[idx];
	}

	vector<int> oldLabel(m,0);
	do {
		printf("step = %d\n", nSteps);
		for(int i = 0; i < m; ++i)
			oldLabel[i] = label[i];
		for(int i = 0; i < k; ++i)
			cnt[i] = 0;

		for(int i = 0; i < m; ++i)
		{
			label[i] = assign(data[i]);
		}

		for(int i = 0; i < k; ++i)
			for(int j = 0; j < n; ++j)
				centroids[i][j] = 0.0;

		for(int i = 0; i < m; ++i)
		{
			int c = label[i];
			for(int j = 0; j < n; ++j)
				centroids[c][j] += data[i][j];
			++cnt[c];
		}
		for(int i = 0; i < k; ++i)
			for(int j = 0; j < n; ++j)
				centroids[i][j] /= (double) cnt[i];

		printResult();
		cout << "\n";
		++nSteps;
	} while (!isConverge(oldLabel));
}

void KMean::printResult()
{
	printf("Centroids:\n");
	for(int i = 0; i < k; ++i)
	{
		printf("C[%d]: (", i);
		for(int j = 0; j < n; ++j)
			printf("%0.3f ", centroids[i][j]);
		printf(")\n");
	}
	
	printf("Label:\n");
	for(int i = 0; i < m; ++i)
		printf("%d\n", label[i]);
	
	printf("Cnt:\n");
	for(int i = 0; i < k; ++i)
		printf("%d\n", cnt[i]);
}

void KMean::printResult1()
{
	for(int i = 0; i < m; ++i)
		printf("%d,%d,%d\n", (int) data[i][0], (int) data[i][1], label[i]);
}

KMean::~KMean()
{
	data.clear();
	centroids.clear();
	label.clear();
}

vector<string> tokenize(char* s)
{
	vector<string> v;
	char* token;
	token = strtok(s, DELIMITER);
	while (token)
	{
		string tmp(token);
		v.push_back(tmp);
		token = strtok(NULL, DELIMITER);
	}
	v[v.size() - 1].pop_back();
	return v;
}

double euclidean(vector<double> p1, vector<double> p2)
{
    assert(p1.size() == p2.size());
    double res = 0.0;
    for(int i = 0; i < p1.size(); ++i)
        res += pow(p1[i] -= p2[i], 2.0);
    return sqrt(res);
}

int argmin(vector<double> v)
{
	int Argmin = 0;
	for(int i = 1; i < v.size(); ++i)
		if (v[i] < v[Argmin])
			Argmin = i;
	return Argmin;
}
// End of Implementation
